test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +523,668 @@ func TestRSSelectorImmutability(t *testing.T) {
+			t.Fatalf(".status.replicas %d is not equal to .spec.replicas %d", replicaset.Status.Replicas, *replicaset.Spec.Replicas)
+		}
+	}
+}
+
+func TestPodOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := c.Core().Pods(ns.Name).List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 1)
+	}
+
+	pod := &pods.Items[0]
+	newLabelMap := map[string]string{"new-foo": "new-bar"}
+	podClient := c.Core().Pods(ns.Name)
+
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		pod.Labels = newLabelMap
+		pod, err = c.Core().Pods(ns.Name).Update(pod)
+		if err != nil {
+			return fmt.Errorf("failed to update pod %s: %v", pod.Name, err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if metav1.GetControllerOf(pod) == nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Adoption: RS should add ControllerRef to a pod when the pod's labels change to match its labels
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// Revert to original labels so that the RS can adopt the pod again
+		pod.Labels = labelMap
+		pod, err = c.Core().Pods(ns.Name).Update(pod)
+		if err != nil {
+			return fmt.Errorf("failed to update pod %s: %v", pod.Name, err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		pod, err := podClient.Get(pod.Name, metav1.GetOptions{})
```

Comment:
It's unclear whether this `pod` value is meant to escape from this callback or not. To make it clear, rename this value from Get to `newPod`. Then, if `GetControllerOf(newPod)` passes the condition, set `pod = newPod` before returning.

Reply:
Yup, this will be clearer.